test/solver/solver.cpp

```diff
@@ -0,0 +1,979 @@
+#include <ginkgo/core/solver/cb_gmres.hpp>
+#include <ginkgo/core/solver/cg.hpp>
+#include <ginkgo/core/solver/cgs.hpp>
+#include <ginkgo/core/solver/fcg.hpp>
+#include <ginkgo/core/solver/gmres.hpp>
+#include <ginkgo/core/solver/idr.hpp>
+#include <ginkgo/core/solver/ir.hpp>
+#include <ginkgo/core/solver/lower_trs.hpp>
+#include <ginkgo/core/solver/upper_trs.hpp>
+
+
+#include "core/test/utils.hpp"
+#include "test/utils/executor.hpp"
+
+
+#if GINKGO_DPCPP_SINGLE_MODE
+using solver_value_type = float;
+#else
+using solver_value_type = double;
+#endif  // GINKGO_DPCPP_SINGLE_MODE
+
+
+template <typename SolverType>
+struct SimpleSolverTest {
+    using solver_type = SolverType;
+    using value_type = typename solver_type::value_type;
+    using index_type = gko::int32;
+    using matrix_type = gko::matrix::Csr<value_type, index_type>;
+    using precond_type = gko::preconditioner::Jacobi<value_type, index_type>;
+
+    static bool is_iterative() { return true; }
+
+    static bool is_preconditionable() { return true; }
+
+    static double tolerance() { return 1e3 * r<value_type>::value; }
+
+    static void preprocess(gko::matrix_data<value_type, index_type>& data)
+    {
+        // make sure the matrix is well-conditioned
+        gko::test::make_hpd(data, 1.5);
+    }
+
+    static typename solver_type::parameters_type build(
+        std::shared_ptr<const gko::Executor> exec,
+        gko::size_type iteration_count)
+    {
+        return solver_type::build().with_criteria(
+            gko::stop::Iteration::build()
+                .with_max_iters(iteration_count)
+                .on(exec));
+    }
+
+    static typename solver_type::parameters_type build_preconditioned(
+        std::shared_ptr<const gko::Executor> exec,
+        gko::size_type iteration_count)
+    {
+        return solver_type::build()
+            .with_criteria(gko::stop::Iteration::build()
+                               .with_max_iters(iteration_count)
+                               .on(exec))
+            .with_preconditioner(
+                precond_type::build().with_max_block_size(1u).on(exec));
+    }
+
+    static const gko::LinOp* get_preconditioner(const solver_type* solver)
+    {
+        return solver->get_preconditioner().get();
+    }
+
+    static const gko::stop::CriterionFactory* get_stop_criterion_factory(
+        const solver_type* solver)
+    {
+        return solver->get_stop_criterion_factory().get();
+    }
+
+    static void assert_empty_state(const solver_type* mtx)
+    {
+        ASSERT_FALSE(mtx->get_size());
+        ASSERT_EQ(mtx->get_system_matrix(), nullptr);
+        ASSERT_EQ(mtx->get_preconditioner(), nullptr);
+        ASSERT_EQ(mtx->get_stopping_criterion_factory(), nullptr);
+    }
+};
+
+
+struct Cg : SimpleSolverTest<gko::solver::Cg<solver_value_type>> {};
+
+
+struct Cgs : SimpleSolverTest<gko::solver::Cgs<solver_value_type>> {
+    static double tolerance() { return 1e5 * r<value_type>::value; }
```

Comment:
Isn't this tolerance a bit too high for `float` ? It might make sense to have a `value_type` specific tolerance ? Something already in the `r<>` struct to get something like `r<value_type>::tolerance` ?

Reply:
Yes, this is kind of the issue in #973, there are some significant differences between Reference and other kernels which I didn't have time to investigate yet